src/Microsoft.OData.Core/JsonLight/ODataJsonLightOutputContext.cs

```diff
@@ -85,43 +72,33 @@ internal sealed class ODataJsonLightOutputContext : ODataOutputContext
             {
                 this.messageOutputStream = messageInfo.MessageStream;
 
-                
-
                 bool isIeee754Compatible = messageInfo.MediaType.HasIeee754CompatibleSetToTrue();
 
-                // We first attempt to create a JSON writer which directly writes to the stream
-                IStreamBasedJsonWriterFactory streamBasedJsonWriterFactory = this.Container?.GetService<IStreamBasedJsonWriterFactory>();
+                IJsonWriterFactory jsonWriterFactory = this.Container?.GetService<IJsonWriterFactory>();
 
-                if (streamBasedJsonWriterFactory != null)
+#if NETCOREAPP
+                if (jsonWriterFactory is ODataUtf8JsonWriterFactory)
                 {
-                    // When using IStreamBasedJsonWriterFactory, we do not wrap the ouput stream in a buffered stream.
+                    // When using ODataUtf8JsonWriterFactory, we do not wrap the ouput stream in a buffered stream.
                     // The default ODataUtf8JsonWriter will be used, and that writer handles
                     // its own buffering (more efficiently). It leads to too many byte[] array allocations if we have
                     // both the BufferedStream and ODataUtf8JsonWriter allocating buffers unnecessarily.
                     this.asynchronousOutputStream = this.messageOutputStream;
                     this.jsonWriter = CreateJsonWriter(
-                        streamBasedJsonWriterFactory,
+                        this.Container,
                         this.messageOutputStream,
                         isIeee754Compatible,
+                        messageWriterSettings,
                         messageInfo.Encoding);
                 }
+#endif
 
-                // Then fallback to the TextWriter-based approach
+                // Fallback if not using ODataUtf8JsonWriterFactory
                 if (this.jsonWriter == null)
                 {
-                    Stream outputStream;
-                    if (this.Synchronous)
-                    {
-                        outputStream = this.messageOutputStream;
-                    }
-                    else
-                    {
-                        this.asynchronousOutputStream = new BufferedStream(this.messageOutputStream, this.MessageWriterSettings.BufferSize);
-                        outputStream = this.asynchronousOutputStream;
-                    }
+                    Stream outputStream = InitializeOutputStream();

```

Comment:
Since both factories now accept a stream, I think the code would look cleaner if the `DefaultJsonWriterFactory` took encapsulated the buffered stream and wrapped it around the stream it received before passing it to the `JsonWriter`. That way you don't have don't have to explicitly check which type of `jsonWriterFactory` we have here. Ideally those details would be hidden within each `IJsonWriterFactory` implementation.

Reply:
@habbes The challenge is as you [noted here](https://github.com/OData/odata.net/pull/2874#discussion_r1502062328). I didn't feel it'd be good design to introduce a parameter for passing the synchronous/asynchronous value. One thing I'd like to look into deeper into is whether we could use the buffered stream in both synchronous and asynchronous scenarios. But I wouldn't wish to expand the scope for this work item for now. I'll create a work item to track the follow up work.

Reply:
@habbes https://github.com/OData/odata.net/issues/2882